Disallow direct item access of NotRequired TypedDict properties: …

bsmedberg-xometry opened the pull request:
### Description

Disallow direct item access of NotRequired TypedDict properties: these should

always be accessed through .get() because the keys may not be present.

Fixes #12094

## Test Plan

Includes unit test coverage for both item access (__getitem__ as an rvalue) and item setting (__setitem__ as an lvalue).

Two existing tests required updates because they partially exercized the now-forbidden codepath using `reveal_type`.

JelleZijlstra commented:
cc @davidfstr

bsmedberg-xometry commented:
I hadn't run the mypyc tests before submitting because they weren't part of the default `runtests.py` setup. I can reproduce the error locally but it seems that mypyc has an unexpected behavior. In mypyc it seems that a typeddict doesn't have a .get() method? At least when I tried to replace d3["c"] with d3.get("c") I get an error back: `error: "TD3" has no attribute "get"`

bsmedberg-xometry commented:
I will come back to this next week to fix .get(), but I think I also need advice on whether this will adversely affect the python ecosystem.

bsmedberg-xometry commented:
@JukkaL I see that you reviewed and approved the previous work on this at #10370 and before that related #9906 - please let me know if you're the right person to review or provide guidance, now that I have the automated tests passing 100%

thomkeh commented:
One of the errors in https://github.com/Gobot1234/steam.py is on this line:

```python
self.colour = int(data["name_color"], 16) if "name_color" in data else None
```
but surely that is valid?

bsmedberg-xometry commented:
@davidfstr roger, understood.

The new commits I've just added implement the narrowing logic (with a test). I've also updated the documentation to match the new proposed behavior and document the more ergonomic NotRequired method of constructing TypedDict with mixed required/not-required elements.

davidfstr commented on mypy/plugins/default.py:
```diff
@@ -238,37 +238,65 @@ def typed_dict_get_signature_callback(ctx: MethodSigContext) -> CallableType:
+    for key in keys:
+        value_type = get_proper_type(ctx.type.items.get(key))
+        if value_type is None:
+            # It would be nice to issue a "TypedDict has no key {key}" failure here. However,
+            # we don't do this because in the case where you have a union of typeddicts, and
+            # one of them has the key but others don't, an error message is incorrect, and
+            # the plugin API has no mechanism to distinguish these cases.
+            output_types.append(default_type or NoneType())
+            continue
+
+        if ctx.type.is_required(key):
+            # Without unions we could issue an error for .get('required_key', default) because
+            # the default doesn't make sense. But because of unions, we don't do that.
+            output_types.append(value_type)
+            continue
```
This looks a bit unusual to me. However I am unfamilar with mypy's plugin system - used by this file - so can't comment on the workaround.

**Someone else familar with mypy plugins may want to look over this code more carefully.**

davidfstr replied:
@JukkaL , it looks like you've made all previous notable modifications to `typed_dict_get_callback` and may have designed the plugin system.

Could you take a look at this code yourself, or suggest a different reviewer who would be familiar with the limits of mypy's plugin system?

davidfstr replied:
Did finish my own review of `typed_dict_get_callback` here. Further review (by JukkaL or others) is no longer requested.

davidfstr commented on test-data/unit/check-literal.test:
```diff
@@ -2222,12 +2222,14 @@ c_key: Literal["c"]
 reveal_type(d[a_key])         # N: Revealed type is "builtins.int"
-reveal_type(d[b_key])         # N: Revealed type is "builtins.str"
+reveal_type(d[b_key])         # N: Revealed type is "builtins.str" \
+                              # E: TypedDict "Outer" key "b" is not required.
+reveal_type(d.get(b_key))     # N: Revealed type is "builtins.str"
 d[c_key]                      # E: TypedDict "Outer" has no key "c"
 
-reveal_type(d.get(a_key, u))  # N: Revealed type is "Union[builtins.int, __main__.Unrelated]"
+reveal_type(d.get(a_key, u))  # N: Revealed type is "builtins.int"
 reveal_type(d.get(b_key, u))  # N: Revealed type is "Union[builtins.str, __main__.Unrelated]"
-reveal_type(d.get(c_key, u))  # N: Revealed type is "builtins.object"
+reveal_type(d.get(c_key, u))  # N: Revealed type is "__main__.Unrelated"
```
Nice to see better inference happening here. 🎉 (No changes requested.)

davidfstr reviewed (commented):
Exciting feature! The types revealed in the test suite are now a lot more precise. 👌 

* Did leave a few rewording suggestions.
* There is a mypy plugin change I marked as requesting review from someone else more-familiar with plugins.

davidfstr commented on docs/source/more_types.rst:
```diff
@@ -1120,18 +1118,15 @@ Now ``BookBasedMovie`` has keys ``name``, ``year`` and ``based_on``.
-you to mix required and non-required (using ``total=False``) items
-in a single ``TypedDict``. Example:
+When a ``TypedDict`` has a mix of items that are required and not required,
+the ``NotRequired`` type annotation can be used to specify this for each field:
 
 .. code-block:: python
 
-   class MovieBase(TypedDict):
+   class Movie(TypedDict):
        name: str
        year: int
-
-   class Movie(MovieBase, total=False):
-       based_on: str
+       based_on: NotRequired[str]
```
This documentation is so much more clear with this change. 👍 (No changes requested.)

davidfstr commented on mypy/checker.py:
```diff
@@ -4865,6 +4860,56 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
             expr = parent_expr
             expr_type = output[parent_expr] = make_simplified_union(new_parent_types)
 
+    def refine_optional_in(self,
```
The narrowing support provided by this commit is very cool, but I don't have enough brainpower this evening to look at all the new code in this commit in detail.

**Requesting a second pair of eyes on the narrowing logic in this commit.**

davidfstr replied:
Did review the narrowing support here. No further review is requested.

davidfstr reviewed (commented):
> The new commits I've just added implement the narrowing logic (with a test). I've also updated the documentation to match the new proposed behavior and document the more ergonomic NotRequired method of constructing TypedDict with mixed required/not-required elements.

Very cool. Did review these new commits as well.

bsmedberg-xometry commented on mypy/messages.py:
```diff
@@ -1276,6 +1276,17 @@ def typeddict_key_not_found(
+        type_name: str = ""
+        if not typ.is_anonymous():
+            type_name = format_type(typ) + " "
+        self.fail('TypedDict {}key "{}" is not required.'.format(
```
Updated the message in the most recent commits.

davidfstr commented:
> Diff from [mypy_primer](https://github.com/hauntsaninja/mypy_primer), showing the effect of this PR on open source code: [...]

(I really like the new error message. 😁)

I'm still planning to look at the new narrowing logic in [mypy/checker.py](https://github.com/python/mypy/pull/12095/files/37f9e7b687532634bc02677234422139a2ec42e8..04b260ac132d7a377d55ad3fe2d6822d870f5208#diff-f96a2d6138bc6cdf2a07c4d37f6071cc25c1631afc107e277a28d5b59fc0ef04) a little later this week/weekend, once I find a larger block of time.

bsmedberg-xometry commented:
When using this PR against our production codebase, I've discovered an issue with this PR, specifically about the narrowing code. This is an example testcase that now fails but shouldn't:

```python
from typing import TypedDict, NotRequired, Optional
Foo = TypedDict('Foo', {'complete': NotRequired[bool]})

def get_if_complete(foo: Foo) -> Foo:
    if "complete" not in foo:
        foo["complete"] = False
        return foo

    return foo
```

This testcase fails with the message `error: Incompatible return value type (got "Foo", expected "Foo")`

1. This is because one is `Foo({a?: bool})` and the other is narrowed to `Foo({a: bool})`
2. And according to the current subtyping rules, a typeddict with a required field is not a subtype of a typeddict with the same optional field. The code at https://github.com/python/mypy/blob/872c03da11b7e8a11c724e337577b2233573b3b3/mypy/subtypes.py#L406 explains the current logic, implemented by @JukkaL many years ago.

I personally disagree with the original decision-making here. Ideally I'd like to allow structural subtyping like this:

```python
class TD1(TypedDict):
    a: str
    b: int

class TD2(TypedDict):
    a: Optional[str]
    b: NotRequired[int]

td1: TD1
td2: TD2 = td1
```

But this does somewhat alter the ergonomics and safety guarantees of the type system: because you could then do things like this:

```python
del td2["b"]
td2["a"] = None
```

And because td2 is the same object as td1, you'd be making changes against the type contract of TD1. In most structural-subtyping systems (including typescript), you're allowed to convert from a more-specific type to the less-specific type, with the understanding that this could allow type-incorrect mutations to occur.

This is a trivially-easy change to make in subtypes.py, and I have a commit ready for it, but I'd like to consider separating that into another PR (which would land first, before this one) since that could overcomplicate this one.

JukkaL commented:
Thanks for the PR! Since I'm going to focus on preparing the 0.940 release, I probably won't be able to review this very soon (sorry!).

To other reviewers: Please don't merge this PR before I've accepted this. This change might be disruptive, but I'll need to investigate it further to get a good estimate of the impact.

davidfstr commented on mypy/checker.py:
```diff
@@ -4865,6 +4860,56 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
+                           collection_type: Type,
+                           ) -> Optional[Type]:
+        """
+        Check whether a condition `optional_item in collection_type` can narrow away Optional.
```
Probably you meant:

> Check whether a condition "**item_type** in collection_type" [...]

(There is no `optional_item` parameter for this docstring to refer to)

davidfstr commented on mypy/checker.py:
```diff
@@ -4609,34 +4610,28 @@ def has_no_custom_eq_checks(t: Type) -> bool:
+                        if narrowed_left_type:
+                            if_map = {operands[left_index]: narrowed_left_type}
+
+                    elif right_is_narrowable:
```
If `left_is_narrowable`, shouldn't we still try to narrow the right side too (if `right_is_narrowable`)?

In particular should this just be `if` rather than `elif`?

davidfstr commented on mypy/checker.py:
```diff
@@ -4865,6 +4860,56 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
+                            collection_type: Type,
+                            ) -> Optional[Type]:
+        """
+        Check whether a condition `'literal' in typeddict` can narrow a non-required ite
```
Nit: "ite" -> "item"

davidfstr reviewed (commented):
Did review [mypy/checker.py](https://github.com/python/mypy/pull/12095/files#diff-f96a2d6138bc6cdf2a07c4d37f6071cc25c1631afc107e277a28d5b59fc0ef04), leaving a few nit comments.

davidfstr commented on mypy/plugins/default.py:
```diff
@@ -238,37 +238,65 @@ def typed_dict_get_signature_callback(ctx: MethodSigContext) -> CallableType:
+
+    default_type: Optional[Type]
+    if len(ctx.arg_types) == 1:
+        default_type = None
```
Seems like you could simplify by making this `default_type = NoneType()` here:
* You can then simplify the type of `default_type: Optional[Type]` to `default_type: Type`.
* You can then simplify the `default_type or NoneType()` expression later to just `default_type`.
* You can (probably) then eliminate the special `if default_type is None:` block.
   * However after eliminating that block you might need an extra length check in the following block: `if len(ctx.args[1]) == 1:`, such as `if len(ctx.args) >= 2 and len(ctx.args[1]) == 1`.

davidfstr reviewed (commented):
Did finish reviewing [mypy/plugins/default.py](https://github.com/python/mypy/pull/12095/files/37f9e7b687532634bc02677234422139a2ec42e8#diff-75a1138400ce78b0cfa3c667429ec71990ff9f30b073fe3fc8207a6d8ccdfad4) . Did propose a simplification of the current logic.

I have now given feedback on all files originally in this PR.

bsmedberg-xometry commented:
I'm going to get back to this at the end of the week. Expectations:

1. I'm going to remove the narrowing code. It is over-aggressive: it really should be value narrowing not type narrowing, and so it prevents a bunch of perfectly valid code (and also cannot land without #12142 which isn't going to be done soon).
2. I'm not yet sure about the `mypy/plugins/default.py` suggestion because I need to understand how it will affect the other bits of the function, but will try.
3. I'll fix up the other review comments, which I understand are are clearly wins.

davidfstr commented:
@bsmedberg-xometry , did life get busy? ^_^

davidfstr commented:
Would it be helpful or welcome @bsmedberg-xometry if I helped apply feedback on this branch?

The branch's diff does implement (Not)Required[] docs in mypy, which is the last piece of the (Not)Required[] functionality that is missing for mypy & CPython 3.12. I'm actively finishing (Not)Required[] functionality and so have active energy to spare toward completing this PR.

bsmedberg-xometry commented:
@davidfstr et al: yes life and work got too busy and I dropped this. I would be happy for somebody to take this over. I recommend still as I had before: ripping out the narrowing code because #12142 was rejected (fundamentally), and that cannot work without the other. That may require some other layers of fixup.

bsmedberg-xometry commented:
I'm going to close this PR and I've opened up #14717 which unbitrots this and does the changes I've mentioned above.
